OpenDreamShared/Compiler/DMF/DMFParser.cs

```diff
@@ -181,54 +196,60 @@ public class DMFParser : Parser<char> {
             return null;
         }
 
-        public MenuElementDescriptor MenuElement() {
+        public MenuElementDescriptor MenuElement(ISerializationManager serializationManager) {
             if (Check(TokenType.DMF_Elem)) {
                 Token elementNameToken = Current();
                 bool hasName = Check(TokenType.DMF_String);
                 Newline();
 
-                MenuElementDescriptor descriptor = new MenuElementDescriptor(hasName ? (string)elementNameToken.Value : null);
-                Dictionary<string, Token> attributes = Attributes();
-
-                SetAttributes(descriptor, attributes);
+                var attributes = Attributes();
+                if (hasName) attributes.Add("name", (string) elementNameToken.Value);
 
-                return descriptor;
+                return serializationManager.ReadValue<MenuElementDescriptor>(attributes);
             }
 
             return null;
         }
 
-        public (string, Token)? AttributeAssign() {
+        public bool TryGetAttribute([NotNullWhen(true)] out string key, [NotNullWhen(true)] out string token)
+        {
+            key = null;
+            token = null;
             Token attributeToken = Current();
 
             if (Check(new[] { TokenType.DMF_Attribute, TokenType.DMF_Macro, TokenType.DMF_Menu, TokenType.DMF_Stretch, TokenType.DMF_Left, TokenType.DMF_Right })) {
                 if (!Check(TokenType.DMF_Equals)) {
                     ReuseToken(attributeToken);
 
-                    return null;
+                    return false;
                 }
 
                 Token attributeValue = Current();
                 if (!Check(ValidAttributeValueTypes)) Error("Invalid attribute value (" + attributeValue.Text + ")");
 
                 Newline();
-                return (attributeToken.Text, attributeValue);
+                key = attributeToken.Text;
+                if(attributeValue.Type != TokenType.DMF_None) token = attributeValue.Text;
+                return true;
             }
 
-            return null;
+            return false;
         }
 
-        public Dictionary<string, Token> Attributes() {
-            Dictionary<string, Token> attributes = new();
+        public MappingDataNode Attributes()
+        {
+            var node = new MappingDataNode();
 
-            while (true) {
-                (string, Token)? attributeAssign = AttributeAssign();
-
-                if (attributeAssign.HasValue) attributes.Add(attributeAssign.Value.Item1, attributeAssign.Value.Item2);
-                else break;
+            while (TryGetAttribute(out var key, out var value))
+            {
+                if (value[0] == '"')
+                {
+                    value = value.Substring(1, value.Length - 2);
```

Comment:
What if `value` is an unterminated string?

Reply:
...Do unterminated strings not error anyways?